src/indexer.rs

```diff
@@ -318,4 +345,331 @@ where
         let block = self.store.get_block_by_height(height)?;
         Ok(block)
     }
+
+
+}
+
+/// Estimates the fee rate for the next block based on the middle transaction in the provided block.
+///
+/// This function analyzes the fee rate of the middle transaction in a block to provide an estimate
+/// for future fee calculations. It implements a simple heuristic by selecting the transaction at
+/// the median position within the block's transaction list.
+///
+/// # Implementation Details
+///
+/// The function uses Bitcoin Core RPC `getrawtransaction` with verbosity level 2, which requires
+/// Bitcoin Core version 25.0.0 or higher. It calculates the fee rate using the formula:
+/// `fee_rate = transaction_fee_in_sats / transaction_vsize_in_vb`
+///
+/// # Note
+///
+/// This is a simple estimation method that could be improved by:
+/// - Using `getblock` with verbosity 3 (when supported by bitcoincore-rpc)
+/// - Calculating median or average fee rates from multiple transactions
+/// - Considering block fullness percentage for more accurate predictions
+fn estimate_fee_rate<B: BitcoinClientApi>(bitcoin_client: &B, next_block: &BlockInfo) -> Result<u64, IndexerError> {
+
+        // Const might be settings in the future
+        const MIN_BLOCK_TX: usize = 5;
+        const ERROR_FEE_RATE: u64 = 0; // sat/vB
+        const DEFAULT_MIN_FEE_RATE: u64 = 1; // sat/vB
+
+        // TODO use get block with verbosity 3 would be a nice optimization (not yet supported by bitcoincore-rpc),
+        // it will remove the need to call again getrawtransaction
+        // with that approach we can cheaply calculate the median fee rate of the block, or the average fee rate from the medium zone transactions
+        // See https://bitcoincore.org/en/doc/23.0.0/rpc/blockchain/getblock/
+
+        let block_tx_count = next_block.txs.len();
+        info!("Transactions count: {}", block_tx_count);
+
+        // In a future version we can analyze if the block is full or empty or which %,
+        // a block lower than 75% will lead to lower fee rates for next block inclusion
+        if block_tx_count <= MIN_BLOCK_TX {
+            warn!("Can't estimate fee rate - block has {} or fewer transactions", MIN_BLOCK_TX);
+            return Ok(ERROR_FEE_RATE);
+        }
+
+        let middle_index = block_tx_count / 2;
+        let middle_tx = next_block.txs[middle_index].clone();
+
+        // Note: For coinbase transactions (first tx in block), there's no fee since there are no inputs
+        // Usually the block is ordered with coinbase as first transaction, but this is not enforced by consensus rules, to this case might rarely happen
+        if middle_tx.is_coinbase() {
+            warn!("Can't estimate fee rate - middle transaction is coinbase");
```

Comment:
This will likely flood the regtest logs with warnings. Can we downgrade it to trace? It doesn’t seem important enough to warrant a warning.

Reply:
this will almost never happen, its a very weird case to have the coinbase tx in the middle of the block, usually is the 1st tx